### PR TITLE
[SYCL][ESIMD]Refactor several math functions and remove duplicate constants

### PR DESCRIPTION
**Pull request by fineg74**
Complementary test PR: https://github.com/intel/llvm-test-suite/pull/1405

### CODE REVIEW

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp [hunk 1]
```diff
@@ -1152,7 +1152,7 @@ sincos(__ESIMD_NS::simd<float, SZ> &dstcos, U src0, Sat sat = {}) {
 
 /// @cond ESIMD_DETAIL
 namespace detail {
-constexpr double HDR_CONST_PI = 3.1415926535897932384626433832795;
+constexpr double __HDR_CONST_PI = 3.1415926535897932384626433832795;
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1155]:
```suggestion
constexpr double __ESIMD_CONST_PI = 3.1415926535897932384626433832795;
```

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp [hunk 2]
```diff
@@ -1162,9 +1162,13 @@ ESIMD_NODEBUG ESIMD_INLINE
     atan(__ESIMD_NS::simd<T, SZ> src0) {
   __ESIMD_NS::simd<T, SZ> Src0 = __ESIMD_NS::abs(src0);
 
-  __ESIMD_NS::simd_mask<SZ> Neg = src0 < T(0.0);
+  __ESIMD_NS::simd<float, SZ> OneP(1.f);
+  __ESIMD_NS::simd<float, SZ> OneN(-1.f);
+  __ESIMD_NS::simd<float, SZ> sign;
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1167]:
```suggestion
  __ESIMD_NS::simd<T, SZ> OneP((T)1.f);
  __ESIMD_NS::simd<T, SZ> OneN((T)-1.f);
  __ESIMD_NS::simd<T, SZ> sign;
```

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp [hunk 3]
```diff
@@ -1179,9 +1183,9 @@ ESIMD_NODEBUG ESIMD_INLINE
        ((Src0 * T(0.395889) + T(1.12158)) * Src0P2) + (Src0 * T(0.636918)) +
        T(1.0));
 
-  Result.merge(Result - T(detail::HDR_CONST_PI / 2.0), Gt1);
-  Result.merge(Result, Neg);
-  return Result;
+  Result.merge(Result - T(detail::__HDR_CONST_PI / 2.0), Gt1);
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1186]:
```suggestion
  Result.merge(Result - T(detail::__ESIMD_CONST_PI / T(2.0)), Gt1);
```

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp [hunk 4]
```diff
@@ -1309,41 +1313,29 @@ template <int N> __ESIMD_NS::simd<float, N> tanh(__ESIMD_NS::simd<float, N> x);
...
-
-  a1.merge(detail::CONST_PI, x < 0.0f);
-
-  __ESIMD_NS::simd<float, N> xy = x * y;
-  __ESIMD_NS::simd<float, N> x2 = x * x;
-  __ESIMD_NS::simd<float, N> y2 = y * y;
-
   /* smallest such that 1.0+CONST_DBL_EPSILON != 1.0 */
   constexpr auto CONST_DBL_EPSILON = 0.00001f;
+  __ESIMD_NS::simd<float, N> OneP(1.f);
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1322]: similar notes as above

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp [hunk 5]
```diff
@@ -1257,306 +1263,234 @@ __ESIMD_API std::enable_if_t<std::is_floating_point<T>::value, T> asin(T src0) {
...
-                            (OneP + x2 * 0.0238095f *
-                                        (OneN + x2 * 0.0138889f *
-                                                    (OneP - x2 * 0.0090909f))));
-  fTrig *= sign;
-  return fTrig;
+template <typename T, int N>
+ESIMD_INLINE __ESIMD_NS::simd<T, N> cos_emu(__ESIMD_NS::simd<T, N> x) {
+  static_assert(std::is_floating_point<T>::value,
+                "Floating point argument type is expected.");
+  return esimd::sin_emu((T)0.5 * (T)detail::__ESIMD_CONST_PI - x);
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1478]: Is itr a misprint here? Why cos_emu calls sin_emu here?

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp [hunk 6]
```diff
@@ -1257,306 +1263,234 @@ __ESIMD_API std::enable_if_t<std::is_floating_point<T>::value, T> asin(T src0) {
...
+
+  __ESIMD_NS::simd<T, N> sign;
+  __ESIMD_NS::simd<T, N> fTrig;
+  __ESIMD_NS::simd<T, N> TwoPI((T)detail::__ESIMD_CONST_PI * (T)2.0);
+  __ESIMD_NS::simd<T, N> CmpI((T)detail::__ESIMD_CONST_PI);
+  __ESIMD_NS::simd<T, N> OneP((T)1.0);
+  __ESIMD_NS::simd<T, N> OneN((T)-1.0);
 
   x = esimd::fmod(x, TwoPI);
+  x.merge(TwoPI + x, x < 0);
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1445]: Can you please explain this line? I don't see equivalent for it in the code "before" patch.
Does this PR have not only restructuring changes, but also sone math fix?

Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1445]: Ok, from phone talk I understood, this algorithm did not work correctly for some negative values.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp [hunk 7]
```diff
@@ -1257,306 +1263,234 @@ __ESIMD_API std::enable_if_t<std::is_floating_point<T>::value, T> asin(T src0) {
 
 /* atan2_fast - a fast atan2 implementation */
 /* vector input */
-template <int N>
-__ESIMD_NS::simd<float, N> atan2_fast(__ESIMD_NS::simd<float, N> y,
-                                      __ESIMD_NS::simd<float, N> x);
+template <typename T, int N>
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1266]: Not that this is an interface change. Prior to this change atan2_fast always returned float. Is this intentional?
Also applies to other APIs. Please preserve the original return types, unless there is some reason not to.

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**